geopandas/sindex.py

```diff
@@ -1,25 +1,341 @@
+                for i in tree_query:
+                    if self._prepared_geometries[i] is None:
+                        # if not already prepared, prepare and cache
+                        self._prepared_geometries[i] = prep(self._geometries[i])
+                    if getattr(self._prepared_geometries[i], predicate)(geometry):
+                        res.append(i)
+                tree_query = res
+            elif predicate == "contains" and len(tree_query) > 1:
+                # prepare this geometry
+                geometry = prep(geometry)
+                tree_query = [
+                    i
+                    for i in tree_query
+                    if getattr(geometry, predicate)(self._geometries[i])
+                ]
+
+            elif predicate is not None:
+                tree_query = [
+                    i
+                    for i in tree_query
+                    if getattr(geometry, predicate)(self._geometries[i])
+                ]
+
+            if not sort:
+                return np.array(tree_query)  # unsorted
+
+            # sort
+            return np.sort(np.array(tree_query))
+
+        def query_bulk(self, geometry, predicate=None, sort=False):
+            """Compatibility layer for pygeos.query_bulk.
+
+            Iterates over `geometry` and queries index.
+            This operation is not vectorized and may be slow.
+            Use PyGEOS with `query_bulk` for speed.
+
+            Parameters
+            ----------
+            geometry : {GeoSeries, GeometryArray, numpy.array of PyGEOS geometries}
+                Accepts GeoPandas geometry iterables (GeoSeries, GeometryArray)
+                or a numpy array of PyGEOS geometries.
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
+                The predicate to use for testing geometries from the tree
+                that are within the input geometry's envelope.
+                See `pygeos.strtree` documentation for more details on
+                expected behavior of predicates.
+            sort : bool, default False
+                Determines if the results should be sorted lexicographically using
+                geometry's indexes as the primary key and the sindex's indexes as the
+                secondary key.
+
+            Returns
+            -------
+            ndarray with shape (2, n)
+                The first subarray contains input geometry indexes.
+                The second subarray contains tree geometry indexes.
+            """  # noqa: E501
+            # Iterates over geometry, applying func.
+            tree_index = []
+            geo_index = []
+
+            for i, geo in enumerate(geometry):
+                res = self.query(geo, predicate=predicate, sort=sort)
+                if res.size > 0:
+                    # sort results and append
+                    tree_index.extend(res)
+                    geo_index.extend([i] * len(res))
```

Comment:
I don't think usage of `i` is correct here.  It is using the position of the input geometry in the input geometries, but not its index.  I think you have to do a further check if it has an index (e.g., `GeoSeries`) and get the corresponding index value.

Reply:
Resolving because we are indeed trying to use `iloc/integer` indexes here.